libs/go/aealite/protocols/message.go

```diff
@@ -0,0 +1,178 @@
+	"encoding/json"
+	"errors"
+)
+
+type MessageId int
+type Address string
+type Performative string
+
+type ProtocolMessageInterface interface {
+	Sender() Address
+	SetSender(Address) error
+	To() Address
+	SetTo(Address) error
+	MessageId() MessageId
+	DialogueReference() DialogueReference
+	Target() MessageId
+	Performative() Performative
+	Body() map[string]interface{}
+	HasSender() bool
+	HasTo() bool
+	GetField(name string) interface{}
+	//ValidPerformatives() []Performative TODO temporarily removed
+}
+
+type DialogueMessageWrapper struct {
+	to                Address
+	sender            Address
+	dialogueReference DialogueReference
+	messageId         MessageId
+	target            MessageId
+	body              map[string]interface{}
+	//validPerformatives helpers.Set TODO understand how to set this
+}
+
+// InitFromProtobuf initializes a message from a DialogueMessage protobuf message.
+//  It unpacks 'message id', 'target' and 'dialogue reference'; moreover,
+//  it decodes the content as a JSON object.
+//  Returns error if:
+//  - the JSON decoding fails
+//  - the body does not contain the 'performative'
+//  It performs side-effect on the method receiver.
+func (message *DialogueMessageWrapper) InitFromProtobuf(dialogueMessage *DialogueMessage) error {
+	message.messageId = MessageId(dialogueMessage.MessageId)
+	message.target = MessageId(dialogueMessage.Target)
+	message.dialogueReference = DialogueReference{
+		dialogueMessage.DialogueStarterReference,
+		dialogueMessage.DialogueResponderReference,
+	}
+
+	content := dialogueMessage.Content
+	var data map[string]interface{}
+	err := json.Unmarshal(content, &data)
+	if err != nil {
+		return err
+	}
+	if _, ok := message.body["performative"]; ok {
+		return errors.New("'performative' field not set")
+	}
+	performative := message.body["performative"]
+	if _, ok := performative.(Performative); ok {
+		return errors.New("cannot cast performative field")
+	}
+
+	message.body = data
+	return nil
+}
+
+func (message *DialogueMessageWrapper) Sender() Address {
+	return message.sender
+}
+
+func (message *DialogueMessageWrapper) SetSender(newAddress Address) error {
+	if message.sender != "" {
+		return errors.New("'sender' field already set")
+	}
+	message.sender = newAddress
+	return nil
+}
+
+func (message *DialogueMessageWrapper) To() Address {
+	return message.to
+}
+
+func (message *DialogueMessageWrapper) SetTo(newAddress Address) error {
+	if message.to != "" {
+		return errors.New("'to' field already set")
+	}
+	message.to = newAddress
+	return nil
+}
+
+func (message *DialogueMessageWrapper) MessageId() MessageId {
+	return message.messageId
+}
+
+func (message *DialogueMessageWrapper) DialogueReference() DialogueReference {
+	return message.dialogueReference
+}
+
+func (message *DialogueMessageWrapper) Target() MessageId {
+	return message.target
+}
+
+func (message *DialogueMessageWrapper) Performative() Performative {
+	return message.body["performative"].(Performative)
```

Comment:
this is a type assertion in Golang. It might fail, but if we `InitFromProtobuf` then the assert should pass (`body` is not exposed and `Body()` returns a copy of it)